File: src/u2ctl/selectors/parser.py
"""Selector string and flag parsing."""

import re
from typing import Dict, Any, Optional
from u2ctl.errors import UsageError


def parse_selector_args(args: Dict[str, Any]) -> Dict[str, Any]:
    """Parse dedicated selector flags (--text, --text-contains, --resource-id, --description, --desc-contains, --bounds) or raw selector dict."""
    selector = {}

    if args.get("text"):
        selector["text"] = args["text"]
    if args.get("text_contains"):
        selector["text_contains"] = args["text_contains"]
    if args.get("resource_id"):
        selector["resource_id"] = args["resource_id"]
    if args.get("description"):
        selector["description"] = args["description"]
    if args.get("desc_contains"):
        selector["desc_contains"] = args["desc_contains"]
    if args.get("bounds"):
        bounds_str = args["bounds"]
        # Format: X1,Y1-X2,Y2 or [X1,Y1][X2,Y2]
        m = re.match(r"\[?(\d+),\s*(\d+)\]?\[?(\d+),\s*(\d+)\]?", bounds_str.replace("-", "]["))
        if not m:
            raise UsageError(f"Invalid bounds format: '{bounds_str}'. Expected 'X1,Y1-X2,Y2' or '[X1,Y1][X2,Y2]'")
        selector["bounds"] = [int(m.group(1)), int(m.group(2)), int(m.group(3)), int(m.group(4))]

    if not selector and args.get("selector"):
        raw = args["selector"]
        if raw.startswith("text:"):
            selector["text"] = raw[5:]
        elif raw.startswith("textContains:"):
            selector["text_contains"] = raw[13:]
        elif raw.startswith("resourceId:"):
            selector["resource_id"] = raw[11:]
        elif raw.startswith("desc:"):
            selector["description"] = raw[5:]
        elif raw.startswith("descContains:"):
            selector["desc_contains"] = raw[13:]
        elif raw.startswith("bounds:"):
            bounds_str = raw[7:]
            m = re.match(r"\[?(\d+),\s*(\d+)\]?\[?(\d+),\s*(\d+)\]?", bounds_str.replace("-", "]["))
            if not m:
                raise UsageError(f"Invalid bounds format in selector: '{bounds_str}'")
            selector["bounds"] = [int(m.group(1)), int(m.group(2)), int(m.group(3)), int(m.group(4))]
        else:
            # Default fallback to text selector
            selector["text"] = raw

    if not selector:
        raise UsageError("Must provide at least one selector flag: --text, --text-contains, --resource-id, --description, --desc-contains, or --bounds")

    return selector

File: u2bun/src/selectors/parser.ts
import { UsageError } from "../errors";

export interface SelectorQuery {
  ref?: string;
  resourceId?: string;
  contentDesc?: string;
  descContains?: string;
  text?: string;
  textContains?: string;
  className?: string;
  bounds?: { x1: number; y1: number; x2: number; y2: number };
}

export function parseSelectorArgs(args: Record<string, unknown>): SelectorQuery {
  const query: SelectorQuery = {};

  if (typeof args.ref === "string" && args.ref.trim()) {
    let refVal = args.ref.trim();
    if (!refVal.startsWith("@")) {
      refVal = "@" + refVal;
    }
    query.ref = refVal;
  }
  if (typeof args.resource_id === "string" && args.resource_id.trim()) {
    query.resourceId = args.resource_id.trim();
  }
  if (typeof args.resourceId === "string" && args.resourceId.trim()) {
    query.resourceId = args.resourceId.trim();
  }
  if (typeof args.description === "string" && args.description.trim()) {
    query.contentDesc = args.description.trim();
  }
  if (typeof args.content_desc === "string" && args.content_desc.trim()) {
    query.contentDesc = args.content_desc.trim();
  }
  if (typeof args.desc_contains === "string" && args.desc_contains.trim()) {
    query.descContains = args.desc_contains.trim();
  }
  if (typeof args.descContains === "string" && args.descContains.trim()) {
    query.descContains = args.descContains.trim();
  }
  if (typeof args.text === "string" && args.text.trim()) {
    query.text = args.text.trim();
  }
  if (typeof args.text_contains === "string" && args.text_contains.trim()) {
    query.textContains = args.text_contains.trim();
  }
  if (typeof args.textContains === "string" && args.textContains.trim()) {
    query.textContains = args.textContains.trim();
  }
  if (typeof args.class_name === "string" && args.class_name.trim()) {
    query.className = args.class_name.trim();
  }
  if (typeof args.className === "string" && args.className.trim()) {
    query.className = args.className.trim();
  }

  if (typeof args.bounds === "string" && args.bounds.trim()) {
    const rect = parseBoundsRect(args.bounds.trim());
    if (!rect) {
      throw new UsageError(`Invalid bounds format: '${args.bounds}'. Expected '[x1,y1][x2,y2]' or 'x1,y1,x2,y2'`);
    }
    query.bounds = rect;
  }

  if (
    !query.ref &&
    !query.resourceId &&
    !query.contentDesc &&
    !query.descContains &&
    !query.text &&
    !query.textContains &&
    !query.className &&
    !query.bounds
  ) {
    throw new UsageError("Must specify at least one selector criteria (--ref, --text, --resource-id, --description, --bounds)");
  }

  return query;
}

export function parseBoundsRect(boundsStr: string): { x1: number; y1: number; x2: number; y2: number } | null {
  // Format 1: [270,1754][450,2058]
  const matchBracket = boundsStr.match(/^\[(\d+),\s*(\d+)\]\[(\d+),\s*(\d+)\]$/);
  if (matchBracket) {
    return {
      x1: parseInt(matchBracket[1], 10),
      y1: parseInt(matchBracket[2], 10),
      x2: parseInt(matchBracket[3], 10),
      y2: parseInt(matchBracket[4], 10),
    };
  }

  // Format 2: 270,1754,450,2058 or 270,1754-450,2058
  const matchComma = boundsStr.match(/^(\d+)[\s,]+(\d+)[\s,\-]+(\d+)[\s,]+(\d+)$/);
  if (matchComma) {
    return {
      x1: parseInt(matchComma[1], 10),
      y1: parseInt(matchComma[2], 10),
      x2: parseInt(matchComma[3], 10),
      y2: parseInt(matchComma[4], 10),
    };
  }

  return null;
}
